File: androidApp/src/main/kotlin/se/soderbjorn/lunamux/android/ui/AboutMenu.kt
/**
 * Top-bar "About & links" overflow menu for the Lunamux Android app.
 *
 * This file contains [AboutMenu], the single info-button-plus-dropdown control
 * shared by the Hosts and Sessions top app bars. It gathers the app's external
 * links (community support forum, marketing website, and the legal pages) into
 * one consistent place on every primary screen, rather than duplicating them
 * inline (and risking drift) per screen.
 */
package se.soderbjorn.lunamux.android.ui

import androidx.compose.foundation.layout.Box
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.HelpOutline
import androidx.compose.material.icons.outlined.Description
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material.icons.outlined.Public
import androidx.compose.material.icons.outlined.Star
import androidx.compose.material.icons.outlined.ThumbUp
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalUriHandler
import se.soderbjorn.lunamux.client.viewmodel.LUNAMUX_DISCUSSIONS_URL
import se.soderbjorn.lunamux.client.viewmodel.LUNAMUX_GITHUB_URL
import se.soderbjorn.lunamux.client.viewmodel.LUNAMUX_PLAY_STORE_URL
import se.soderbjorn.lunamux.client.viewmodel.LUNAMUX_PRIVACY_URL
import se.soderbjorn.lunamux.client.viewmodel.LUNAMUX_SITE_URL
import se.soderbjorn.lunamux.client.viewmodel.LUNAMUX_TERMS_URL

/**
 * An info icon for a top app bar that reveals the app's external links.
 *
 * Placed in the `actions` slot of both the Hosts ([HostsScreen]) and Sessions
 * ([TreeScreen]) top bars, it gives users a single, always-reachable entry
 * point to the support forum, the website, and the legal pages from either
 * primary screen. Tapping the icon opens a [DropdownMenu]; selecting an item
 * opens its URL in the browser via [LocalUriHandler] and dismisses the menu.
 *
 * Keeping this a single composable (rather than inline per screen) means the
 * two bars can never drift out of sync on which links they offer, and any new
 * link is added in exactly one place.
 *
 * @see HostsScreen
 * @see TreeScreen
 * @see LUNAMUX_DISCUSSIONS_URL
 */
@Composable
fun AboutMenu() {
    var expanded by remember { mutableStateOf(false) }
    Box {
        IconButton(onClick = { expanded = true }) {
            Icon(
                Icons.Outlined.Info,
                contentDescription = "About & links",
                tint = SidebarTextPrimary,
            )
        }
        DropdownMenu(
            expanded = expanded,
            onDismissRequest = { expanded = false },
        ) {
            AboutMenuItems(onItemSelected = { expanded = false })
        }
    }
}

/**
 * Emits the app's external-link [DropdownMenuItem]s (support forum, website,
 * GitHub, Play Store, and the legal pages) into the enclosing [DropdownMenu].
 *
 * Factored out of [AboutMenu] so the same link list can be reused inside a
 * *host* menu without duplication — the Sessions ([TreeScreen]) top bar folds
 * these links into its combined overflow ("⋮") menu rather than exposing a
 * separate info button, while [AboutMenu] still wraps them in a standalone
 * info-button-plus-dropdown for the Hosts ([HostsScreen]) bar. Keeping the
 * items in one composable means the two entry points can never drift on which
 * links they offer.
 *
 * Must be called from within a [DropdownMenu] (or another `ColumnScope` menu
 * container). Selecting an item opens its URL via [LocalUriHandler] and then
 * invokes [onItemSelected] so the caller can dismiss its own menu.
 *
 * @param onItemSelected called when any link is tapped, before/after the URL is
 *   opened, so the caller can close the surrounding menu (e.g. flip its
 *   `expanded` state to `false`).
 * @see AboutMenu
 * @see TreeScreen
 */
@Composable
fun AboutMenuItems(onItemSelected: () -> Unit) {
    val uriHandler = LocalUriHandler.current
    DropdownMenuItem(
        text = { Text("Support Forum") },
        leadingIcon = {
            Icon(Icons.AutoMirrored.Filled.HelpOutline, contentDescription = null)
        },
        onClick = {
            onItemSelected()
            uriHandler.openUri(LUNAMUX_DISCUSSIONS_URL)
        },
    )
    DropdownMenuItem(
        text = { Text("Website") },
        leadingIcon = {
            Icon(Icons.Outlined.Public, contentDescription = null)
        },
        onClick = {
            onItemSelected()
            uriHandler.openUri(LUNAMUX_SITE_URL)
        },
    )
    DropdownMenuItem(
        text = { Text("Star on GitHub") },
        leadingIcon = {
            Icon(Icons.Outlined.Star, contentDescription = null)
        },
        onClick = {
            onItemSelected()
            uriHandler.openUri(LUNAMUX_GITHUB_URL)
        },
    )
    DropdownMenuItem(
        text = { Text("Rate on Google Play") },
        leadingIcon = {
            Icon(Icons.Outlined.ThumbUp, contentDescription = null)
        },
        onClick = {
            onItemSelected()
            uriHandler.openUri(LUNAMUX_PLAY_STORE_URL)
        },
    )
    // Separate the actionable/engagement links above from the legal
    // boilerplate below.
    HorizontalDivider()
    DropdownMenuItem(
        text = { Text("Privacy Policy") },
        leadingIcon = {
            Icon(Icons.Outlined.Lock, contentDescription = null)
        },
        onClick = {
            onItemSelected()
            uriHandler.openUri(LUNAMUX_PRIVACY_URL)
        },
    )
    DropdownMenuItem(
        text = { Text("Terms") },
        leadingIcon = {
            Icon(Icons.Outlined.Description, contentDescription = null)
        },
        onClick = {
            onItemSelected()
            uriHandler.openUri(LUNAMUX_TERMS_URL)
        },
    )
}

File: iosApp/iosApp/Views/AboutMenu.swift
//
//  AboutMenu.swift
//  iosApp
//
//  Top-bar "About & links" overflow menu for the Lunamux iOS app.
//
//  Defines `AboutMenu`, the single info-button-plus-menu control shared by the
//  Hosts and Sessions toolbars. It gathers the app's external links (community
//  support forum, marketing website, and the legal pages) into one consistent
//  place on every primary screen, rather than duplicating them inline (and
//  risking drift) per screen. Mirrors the Android `AboutMenu` composable.
//

import SwiftUI
import Client

/// An info-icon toolbar button that reveals the app's external links.
///
/// Placed in the trailing toolbar of both the Hosts (`HostsView`) and Sessions
/// (`TreeView`) screens, it gives users a single, always-reachable entry point
/// to the support forum, the website, and the legal pages from either primary
/// screen. Tapping the icon opens a `Menu`; each row opens its URL in the
/// browser.
///
/// Keeping this a single view (rather than inline per screen) means the two
/// toolbars can never drift out of sync on which links they offer, and any new
/// link is added in exactly one place. Mirrors the Android `AboutMenu`
/// composable; the URLs are read from the shared Kotlin `LUNAMUX_*_URL`
/// constants (in `client`'s `ExternalLinks.kt`) via the generated
/// `ExternalLinksKt` accessor, so every platform points at one canonical
/// source and no URL is hardcoded here.
struct AboutMenu: View {
    var body: some View {
        Menu {
            AboutMenuItems()
        } label: {
            Image(systemName: "info.circle")
                // Same tint as every other bar action — mixed per-icon tints
                // read as a bug on light themes (issue #96).
                .foregroundStyle(Palette.textPrimary)
        }
        .accessibilityLabel("About & links")
    }
}

/// The app's external-link rows (support forum, website, GitHub, App Store, and
/// the legal pages) as bare menu content.
///
/// Factored out of `AboutMenu` so the same links can be embedded inside a *host*
/// menu without duplication — the Sessions (`TreeView`) toolbar folds these
/// links into its combined overflow ("ellipsis") menu rather than exposing a
/// separate info button, while `AboutMenu` still wraps them in a standalone
/// info-button-plus-menu for the Hosts (`HostsView`) toolbar. Keeping the rows
/// in one view means the two entry points can never drift on which links they
/// offer. Must be placed inside a `Menu` (or another menu content builder).
///
/// The URLs are read from the shared Kotlin `LUNAMUX_*_URL` constants (in
/// `client`'s `ExternalLinks.kt`) via the generated `ExternalLinksKt` accessor,
/// so every platform points at one canonical source and no URL is hardcoded.
///
/// - SeeAlso: `AboutMenu`
/// - SeeAlso: `TreeView`
struct AboutMenuItems: View {
    /// GitHub Discussions board used as the community support forum. Read from
    /// the shared Kotlin `LUNAMUX_DISCUSSIONS_URL` constant.
    private let discussionsURL = URL(string: ExternalLinksKt.LUNAMUX_DISCUSSIONS_URL)!

    /// Canonical marketing / download site. Read from the shared Kotlin
    /// `LUNAMUX_SITE_URL` constant.
    private let siteURL = URL(string: ExternalLinksKt.LUNAMUX_SITE_URL)!

    /// Public GitHub repository, offered as a "Star on GitHub" action. Read from
    /// the shared Kotlin `LUNAMUX_GITHUB_URL` constant.
    private let gitHubURL = URL(string: ExternalLinksKt.LUNAMUX_GITHUB_URL)!

    /// Apple App Store listing, offered as a "Rate on the App Store" action.
    /// Read from the shared Kotlin `LUNAMUX_APP_STORE_URL` constant.
    private let appStoreURL = URL(string: ExternalLinksKt.LUNAMUX_APP_STORE_URL)!

    /// Published privacy policy page. Read from the shared Kotlin
    /// `LUNAMUX_PRIVACY_URL` constant.
    private let privacyURL = URL(string: ExternalLinksKt.LUNAMUX_PRIVACY_URL)!

    /// Published terms of service page. Read from the shared Kotlin
    /// `LUNAMUX_TERMS_URL` constant.
    private let termsURL = URL(string: ExternalLinksKt.LUNAMUX_TERMS_URL)!

    var body: some View {
        Link(destination: discussionsURL) {
            Label("Support Forum", systemImage: "questionmark.circle")
        }
        Link(destination: siteURL) {
            Label("Website", systemImage: "globe")
        }
        Link(destination: gitHubURL) {
            Label("Star on GitHub", systemImage: "star")
        }
        Link(destination: appStoreURL) {
            Label("Rate on the App Store", systemImage: "hand.thumbsup")
        }
        // Separate the actionable/engagement links above from the legal
        // boilerplate below.
        Divider()
        Link(destination: privacyURL) {
            Label("Privacy Policy", systemImage: "lock")
        }
        Link(destination: termsURL) {
            Label("Terms", systemImage: "doc.text")
        }
    }
}
